### DISCUSSION TITLE
rustdesk-hbbs.service fail to start.

### DISCUSSION THREAD

**Discussion by necklinux** [rustdesk/rustdesk-server-pro#409] [General]
The client can not connect to server.
I check the status of service :

root@ip-172:~# sudo systemctl status rustdesk-hbbr
● rustdesk-hbbr.service - RustDesk Relay Server
     Loaded: loaded (/etc/systemd/system/rustdesk-hbbr.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2024-10-09 04:22:06 UTC; 56min ago
   Main PID: 451 (hbbr)
      Tasks: 5 (limit: 505)
     Memory: 764.0K
        CPU: 292ms
     CGroup: /system.slice/rustdesk-hbbr.service
             └─451 /usr/bin/hbbr

Oct 09 04:22:06 ip-172-26-12-245 systemd[1]: Started RustDesk Relay Server.
root@ip-172:~# sudo systemctl status rustdesk-hbbs
● rustdesk-hbbs.service - RustDesk Signal Server
     Loaded: loaded (/etc/systemd/system/rustdesk-hbbs.service; enabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Wed 2024-10-09 05:18:27 UTC; 5s ago
    Process: 3351 ExecStart=/usr/bin/hbbs (code=exited, status=1/FAILURE)
   Main PID: 3351 (code=exited, status=1/FAILURE)
        CPU: 6ms

It looks rustdesk-hbbs fail to start.
I check log of  hbbs.error:

root@ip-172:~# tail /var/log/rustdesk-server/hbbs.error
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations
Error: migration 20231221032102 was previously applied but is missing in the resolved migrations

How can i fix this issue?
Thanks

**Comment by rustdesk** [comment]
Have you ever modifed the database manually?

**Reply by rustdesk** [reply]
@fufesou follow it up

Backup database files (`db.sqlite3`, `db.sqlite3-shm`,  `db.sqlite3-wal`) first, refering to 

https://www.reddit.com/r/rust/comments/12z6n77/resolving_previously_applied_but_missing_error_in/

and

https://github.com/launchbadge/sqlx/blob/028084bce3a741e995c3e6c559c6dbb27a62534d/sqlx-cli/src/migrate.rs#L267

**Comment by fufesou** [comment]
@necklinux Hi, I tentatively assume that you have modified the database file. If you have not, please let me know.

You can add the missing data in table `_sqlx_migrations`.

Hope you can execute sqlite statements.

Please:
1. Backup database files (`db.sqlite3`, `db.sqlite3-shm`, `db.sqlite3-wal`) first, in `/var/lib/rustdesk-server`.
2. Add the missing data.

### Add the missing data

#### missing data

```sql
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'8e7e45e4e9cc7088f5ef10017c1610386f938c09b3ba295a4e359558b40ec44f17272982c9b3a8f3ac66043aee44eb2a', 'group info', 1560600, '2024-10-09 07:43:15', 1, 20231221032102);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'fa1f0ef5fc37f3835f4a8d3b58a913fc2a61a4ff112edf0ea8906e07ef1c1eb47ca85fbc314e7348a9f08c4ffd7b9c03', 'user auth2fa', 1424200, '2024-10-09 07:43:15', 1, 20231226082926);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'56624d5d48829e06ab73a49a8d714ef80d8279fe0510dbda8f399a2f2ad68cd265d2d2a484994cdbf2354425bc02eb8a', 'third auth remove credential', 3390300, '2024-10-09 07:43:15', 1, 20240110130211);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'ffd84f351ff3b81a8c5fd942ccb9f7eaf59b7f41f57f102ccb9420ac2521dde713d23b3585a8cc3fb484d9dfb247f813', 'user third auth identifier lowercase', 1159700, '2024-10-09 07:43:15', 1, 20240130142656);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'746a083e78d3ecfa2f27e267c4844592531abd5c035058c86658b7edbc3dedc371878940a6558db4f4fb70273304db11', 'third auth identifier recover if not ldap', 718900, '2024-10-09 07:43:15', 1, 20240206101908);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'395fe1ced994122222af99943f5b41a8f697743ce9272f2d4b99e84a12d39d29b5653c87b3f67485ae42f2404e190677', 'custom client', 939700, '2024-10-09 07:43:15', 1, 20240229024003);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'3aaf31c3559b219cd817dfe5a9afcb7afb86711b5fa960624266f0e8e450530f79048243afa5d34a5ce36599b83b8be3', 'ab', 1812800, '2024-10-09 07:43:15', 1, 20240312015505);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'ef4a08b2a15acfac1a115e046210107670127b2b1fe71260dca3e102098df02ae88b3dd88d0e24f766bc1b2ae514f2b1', 'add last online', 4212100, '2024-10-09 07:43:15', 1, 20240410090154);
insert into `_sqlx_migrations` (`checksum`, `description`, `execution_time`, `installed_on`, `success`, `version`) values (X'ddb8522a387ed9c80539a797952afea4db3430e1b0d35e46847a714aa99674f85a02805a3ce255b4c37e3220863f5aed', 'ab peer info json', 6301000, '2024-10-09 07:43:15', 1, 20240530054855);
```

#### delete the added data

```sql
delete from `_sqlx_migrations` where `version` >= 20231221032102;
```

#### operations

Assume we have fore statements to add. `A`, `B`, `C`, `D`

1. Try to add all missing data. (`A`, `B`, `C`, `D`). Go to step 2
2. Try to start the service. Go to step 3
3. If the service fails to start, delete the added data by running ```delete from `_sqlx_migrations` where `version` >= 20231221032102;```. Then try to add the statement without the last statement. That is, if you have already added (`A`, `B`, `C`, `D`), you should now add (`A`, `B`, `C`), and if you have already added (`A`, `B`, `C`), you should now add (`A`, `B`). Go to step 2.

**Reply by necklinux** [reply]
Thank you so much.

1. I did not modifed the database manually .
2. I download rustdesk-utils , reset my password last month.
3. There is no db.sqlite3-shm, db.sqlite3-wal in /var/lib/rustdesk-server
4. I try to add missing data , all of them get error
    `Error: stepping, UNIQUE constraint failed: _sqlx_migrations.version (19)`
5. I try to restart rustdesk-hbbs --> Fail
6. delete from `_sqlx_migrations` where `version` >= 20231221032102;
7. restart rustdesk-hbbs again --> Running
